src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -150,5 +155,96 @@ public void ConnectHeaders_NewAndExistingHeaders_ShouldCreateNewHeaders()
+            Assert.Equal(beforeKeys.Count, afterKeys.Count);
+            Assert.True(beforeKeys.All(afterKeys.Contains));
+        }
+
+        // Supply some headers and after that supply some more headers
+        // and make sure that PeerTipsByHash is updated (the total amount of items is the same).
+        [Fact]
+        public void ConnectHeaders_SupplyHeadersThenSupplyMore_PeerTipsByHashShouldBeUpdated()
+        {
+            TestContext testContext = new TestContext();
+            ChainedHeaderTree cht = testContext.CreateChainedHeaderTree();
+
+            var chainTip = testContext.ExtendAChain(10);
+            cht.Initialize(chainTip, true);
+            chainTip.BlockDataAvailability = BlockDataAvailabilityState.BlockAvailable;
+            ChainedHeader newChainTip = testContext.ExtendAChain(10, chainTip);
+
+            var listOfExistingHeaders = testContext.ChainedHeaderToList(chainTip, 10);
+            var listOfNewHeaders = testContext.ChainedHeaderToList(newChainTip, 10);
+
+            testContext.ChainStateMock.Setup(s => s.ConsensusTip).Returns(chainTip);
+            chainTip.BlockValidationState = ValidationState.FullyValidated;
+
+            cht.ConnectNewHeaders(1, listOfExistingHeaders);
+
+            var peerIdsByTipHashBefore = new Dictionary<uint256, HashSet<int>>(cht.GetPeerIdsByTipHash());
+            var chainedHeaderTreeHashesBefore = cht.GetChainedHeadersByHash();
+
+            cht.ConnectNewHeaders(1, listOfNewHeaders);
+
+            var peerIdsByTipHashAfter = cht.GetPeerIdsByTipHash();
+            var chainedHeaderTreeHashesAfter = cht.GetChainedHeadersByHash();
+
+            // Same number of entries as we are reassigning hash
+            Assert.True(chainedHeaderTreeHashesBefore.Count == chainedHeaderTreeHashesAfter.Count );
+
+            // Peer's tip hash has changed
+            Assert.True(peerIdsByTipHashBefore.FirstOrDefault(x => x.Value.Contains(1)).Key !=
+                        peerIdsByTipHashAfter.FirstOrDefault(x => x.Value.Contains(1)).Key);
+        }
+
+        // Make sure checkpoints are off - supply some headers,
+        // CHT should return ToDownload array of the same size as the amount of headers.
+        [Fact]
+        public void ConnectHeaders_SupplyHeaders_ToDownloadArraySizeSameAsNumberOfHeaders()
+        {
+            // Setup
+            var ctx = new TestContext();
+            ChainedHeaderTree cht = ctx.CreateChainedHeaderTree();
+            var chainTip = ctx.ExtendAChain(10);
+            cht.Initialize(chainTip, true);
+            chainTip.BlockDataAvailability = BlockDataAvailabilityState.BlockAvailable;
+            ctx.ChainStateMock.Setup(s => s.ConsensusTip).Returns(chainTip);
+            ctx.ConsensusSettings.UseCheckpoints = false;
+            ChainedHeader newChainTip = ctx.ExtendAChain(10, chainTip);
+            List<BlockHeader> listOfNewHeaders = ctx.ChainedHeaderToList(newChainTip, 10);
```

Comment:
Not only here, general comment, you always supply the full chain from the 1st to the last header. Although this is fine in terms of test correctness, the test would actually cover little more if you for example use value 8 instead of 10 here so that we simulate the peer to provide not the full chain but only a part of it. Elsewhere you then could 1st provide just 2 or 3 blocks in the 1st run and then just supply another 5 or maybe 15 blocks so that we don't have every test case repeating the same scenario in terms of which chains are supplied in the relation to how CHT was initialised.

So in general, the otherwise would be, add some variability.